### PR TITLE
Use a shared_mutex for read `Dictionary` operations

### PR DESCRIPTION
**Pull request by Al2Klimov**
This allows multiple parallel read operations resulting in a overall speedup on systems with many cores.

# #9627

| real | user | sys |
|------|------|-----|
| 23m41,989s | 1056m24,028s | 207m46,986s |
| 24m10,306s | 1084m32,750s | 199m45,335s |
| 24m58,666s | 1152m39,816s | 168m35,401s |
| 25m16,643s | 1156m18,189s | 201m20,795s |
| 23m5,677s | 1044m19,698s | 186m54,275s |
| 24m26,794s | 1112m13,336s | 184m30,857s |
| 23m5,215s | 1049m22,950s | 186m35,814s |
| 23m30,688s | 1061m56,622s | 184m29,615s |
| 23m37,115s | 1073m8,501s | 185m36,210s |
| 22m32,282s | 1037m58,185s | 166m31,956s |

# #9627 + this PR

| real | user | sys |
|------|------|-----|
| 21m0,810s | 1116m15,330s | 3m52,515s |
| 21m32,898s | 1150m21,358s | 4m11,779s |
| 20m59,009s | 1115m39,142s | 3m54,868s |
| 21m53,785s | 1171m41,773s | 3m58,269s |
| 21m18,770s | 1128m18,154s | 4m1,864s |
| 21m3,465s | 1124m55,727s | 4m5,976s |
| 21m48,964s | 1165m7,774s | 4m3,180s |
| 20m10,108s | 1077m6,399s | 4m9,552s |
| 20m56,285s | 1118m42,794s | 3m56,123s |
| 20m18,122s | 1084m51,305s | 4m1,666s |

### CODE REVIEW

File: lib/base/dictionary.cpp [hunk 1]
```diff
@@ -263,6 +267,7 @@ String Dictionary::ToString() const
 void Dictionary::Freeze()
 {
 	ObjectLock olock(this);
+	std::unique_lock<std::shared_timed_mutex> lock (m_DataMutex);
 	m_Frozen = true;
 }
```
Comment (julianbrost) [inline lib/base/dictionary.cpp:272]: Why is this needed here?

**Review by julianbrost — COMMENTED** [review]
Surprising impact on performance, I wouldn't have expected so much contention on dicts. But from the change in system CPU time, this looks quite similar to the corresponding change in namespaces.

File: lib/base/dictionary.cpp (hunk 1 above)
Comment (Al2Klimov) [reply lib/base/dictionary.cpp:272]: Wasn't it your idea to double-lock all modificators?

Comment (julianbrost) [reply lib/base/dictionary.cpp:272]: Everything that modifies `m_Data`.

In `Namespace`, the idea of `m_DataMutex` is to only protect only the integrity of the `m_Data` data structure itself (hence the name) and indeed, `Namespace::Freeze()` doesn't acquire `m_DataMutex`.

Here, `m_Frozen` is still fully synchronized using the `ObjectLock` (only write operations read it, these also obtain the `ObjectLock`). In `Namespace`, additionally read operations check `m_Frozen` to disable locking, that's why it was made a `std::atomic<bool>` there.

**Comment by Al2Klimov** [comment]
Apply rules eat vars. Vars are dicts.

**Comment by julianbrost** [comment]
Yes, I just expected less contention as these variables are per object, but apparently there is quite some of it.

**Review by julianbrost — APPROVED** [review]
